[Bullet3] fixes config failures.

mathisloge opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  
  Fixes #19234
 Fixes #19172

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
think so

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
yes
- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
yes

mathisloge committed:
fix bullet3

mathisloge committed:
add-version

mathisloge commented:
tensorflow fails with 
```
You have bazel 4.1.0 installed.
Please downgrade your bazel installation to version 3.99.0 or lower to build TensorFlow! To downgrade: download the installer for the old version (from https://github.com/bazelbuild/bazel/releases) then run the installer.
```

So this should not be a thing on my side.

mathisloge committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into bullet-fix

JonLiu1993 commented on ports/bullet3/vcpkg.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "bullet3",
+  "version-string": "3.17",
```
```suggestion
  "version": "3.17",
```

mathisloge committed:
change version-string to version

mathisloge committed:
overwrite version

bttner commented on ports/bullet3/usage:
```diff
@@ -1,4 +1,6 @@
 find_package(Bullet CONFIG REQUIRED)
+include(${BULLET_USE_FILE})
+
 target_link_libraries(main PRIVATE LinearMath Bullet3Common BulletDynamics BulletSoftBody)
```
I am running Ubuntu 20.04, CMake 3.21.1, and this pull request fixed the problem `Unknown CMake command "set_and_check"`  for me (#19234 and #19172). However, as also mentioned in #19172, the proposed item order in target_link_libraries leads to a compiling error (e.g., undefined reference to ...) when running the make command. One can easily reproduce this error by using _HelloWorld.cpp_ in _bullet3/examples/HelloWorld_ and 
[CMakeLists.txt](https://github.com/microsoft/vcpkg/files/6920032/CMakeLists.txt). Using `target_link_libraries(main PRIVATE BulletSoftBody BulletDynamics BulletCollision Bullet3Common LinearMath)` instead solves the issue, at least for me.

mathisloge replied:
Please see the changed usage file. That will resolve this issue too

mathisloge replied:
```` include(${BULLET_USE_FILE}) ``` after find_package

JackBoosY replied:
This file only add the usage cmake code, so we don't need it.

mathisloge replied:
i don't quite understand your comment @JackBoosY . 
The `include(${BULLET_USE_FILE})` is required after `find_package` because it adds the library paths etc. to the library search dir. `LinearMath` is only a library name. It doesn't have a cmake target from upstream. So this would resolve to something like `LinearMath.a or LinearMath.lib`

mathisloge replied:
so the usage file should definitly be appended to the port since it highlights the needed `include(${BULLET_USE_FILE})`

mathisloge replied:
Just for reference, the BULLET_USE_FILE has the following contents: 
```
add_definitions     ( ${BULLET_DEFINITIONS} )
include_directories ( ${BULLET_INCLUDE_DIRS} )
link_directories    ( ${BULLET_LIBRARY_DIRS} )
```

JackBoosY replied:
> Just for reference, the BULLET_USE_FILE has the following contents: 
```
add_definitions     ( ${BULLET_DEFINITIONS} )
include_directories ( ${BULLET_INCLUDE_DIRS} )
link_directories    ( ${BULLET_LIBRARY_DIRS} )
```
According to this content, the only purpose of this file is to automatically use this port in private projects. This approach will cause the target in the project to be unable to be exported and overlap with the rest of the usage. 
So I don’t think we need them.

mathisloge replied:
This is needed. If you don't use BULLET_USE_FILE it will lead to link errors since `LinearMath` might not be found. It isn't in the library search path for the linker. It all is provided from upstream. Again if you don't use `BULLET_USE_FILE` it will lead to link errors. Or you add the `BULLET_LIBRARY_DIRS` to the library search path yourself.

bttner replied:
As can be seen in the linked file - CMakeLists.txt - in my comment, I put

> `include(${BULLET_USE_FILE})` after find_package

but still run into compiling errors when running make. Only after I set `target_link_libraries(HelloWorld PRIVATE BulletSoftBody BulletDynamics BulletCollision Bullet3Common LinearMath)` my project compiled. Thus, the order in target_link_libraries seems important and can not be set arbitrary. 

Regarding the use of `include(${BULLET_USE_FILE})`. When I exclude that line, I run into linking errors, e.g., `fatal error: btBulletDynamicsCommon.h: No such file or directory`.

Overall, I suggest to simply change `target_link_libraries(main PRIVATE LinearMath Bullet3Common BulletDynamics BulletSoftBody)` to `target_link_libraries(HelloWorld PRIVATE BulletSoftBody BulletDynamics BulletCollision Bullet3Common LinearMath)`. At least in my case, everything works as expected then.

mathisloge replied:
That doesn't work for me in a completly new project. 
Could you confirm that the error persists if you clean your build dir and reconfigure?

Because the order of the libraries don't have any effect for me.

bttner replied:
Ok, probably I am doing something wrong. I just did the following:

1. `$ git clone https://github.com/mathisloge/vcpkg.git`, `$ ./vcpkg/bootstrap-vcpkg.sh` and `$ ./vcpkg/vcpkg install bullet3`
2. `$ mkdir HelloWorld`, copied [HelloWorld.cpp](https://github.com/bulletphysics/bullet3/blob/master/examples/HelloWorld/HelloWorld.cpp) and the mentioned CMakeLists.txt into it.
3. `$ cmake -B [build directory] -S . -DCMAKE_TOOLCHAIN_FILE=[path to vcpkg]/...` works.
4. `$ cmake --build [build directory]` only works when I set `target_link_libraries(main PRIVATE BulletSoftBody BulletDynamics BulletCollision Bullet3Common LinearMath)`. Otherwise, I get a lot of error messages. The first reads `undefined reference to btDefaultCollisionConfiguration::btDefaultCollisionConfiguration(btDefaultCollisionConstructionInfo const&)`

mathisloge replied:
In step one you are changing to the branch `mathisloge:bullet-fix`? 

Are you cleaning your build dir between step 3 and 4? 
Otherwise it looks all correct. Maybe add `--config Debug --target all` to step 4

mathisloge replied:
But I just discovered some other problems regarding the vcpkg build structure. If you want to build for debug, you can't find the LinearMath lib since it is called `LinearMath_Debug` and in a different folder. I think i will need to add some patches or a own target for the lib.

mathisloge replied:
@bttner could you pull the new changes again? I've added the missing export targets. You could also remove the `include(${BULLET_USE_FILE})` and replace it with `target_include_directories(main PRIVATE ${BULLET_INCLUDE_DIRS})` after your `add_executable` call.

bttner replied:
Yes, I changed to the branch `mathisloge:bullet-fix` and therefore am using bullet3[core]:x64-linux -> 3.17#1. Furthermore, in the fourth step I do `--config Debug --target all --clean first`. Despite all my efforts, however, I always run into the same issue. Have you tried the provided example project yourself and it compiles without errors?

mathisloge replied:
have you tried the updates i've pushed? 

The `clean-first` option has some strange side effects. Run `--target clean` run the configure step, and then build without `--clean-first`

mathisloge committed:
add export targets; change usage_file

mathisloge committed:
version

mathisloge commented:
@JackBoosY PR is ready to review. I've added the last remaining targets and updated the usage file. 
This takes the file structure layout of vcpkg into accout (lib or debug/lib). 
I've tested with ubuntu and windows the HelloWorld program of bullet3 without any linkage errors (debug and release)

JackBoosY commented on ports/bullet3/usage:
```diff
@@ -1,4 +1,6 @@
 
 find_package(Bullet CONFIG REQUIRED)
+
 target_link_libraries(main PRIVATE LinearMath Bullet3Common BulletDynamics BulletSoftBody)
+target_include_directories(main PRIVATE ${BULLET_INCLUDE_DIRS})
```
For those content, I'll help you to fix that.

mathisloge replied:
I've added the include dirs to each target and checked if they match with what upstream expects.

mathisloge committed:
add bullet include dirs to target

mathisloge committed:
version

mathisloge committed:
remove line break

mathisloge committed:
version

JackBoosY commented:
@mathisloge  In fact, there are still some other issue:
- LinearMath should be dependent on some other components
- The generated cmake file should be in _share/bullet_

And I'm fixing them now.

JackBoosY commented:
Since I have no permission to push my changes to this branch, please use my changes:
[bullet3.zip](https://github.com/microsoft/vcpkg/files/6930388/bullet3.zip)

mathisloge committed:
apply changes from @JackBoosY

mathisloge committed:
version

mathisloge commented:
@JackBoosY thanks for the fixes. 

Maybe you should get the maintainer role to get access? 
![image](https://user-images.githubusercontent.com/13556116/128164625-ed9ef833-4b2d-4cf5-8963-4f6afa274854.png)

JackBoosY commented on ports/bullet3/cmake-fix.patch:
```diff
@@ -1,283 +1,515 @@
-+			DESTINATION ${BULLET_CONFIG_CMAKE_PATH})
- 	ENDIF (NOT INTERNAL_CREATE_DISTRIBUTABLE_MSVC_PROJECTFILES)
- ENDIF (INSTALL_LIBS)
+diff --git a/BulletConfig.cmake.in b/BulletConfig.cmake.in
```
Can you please change this file's EOL?

mathisloge replied:
have changed to LF

mathisloge committed:
change EOL to LF

mathisloge committed:
version

JonLiu1993 commented:
Feature multithreading tested successfully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

mathisloge commented:
will someone merge this?

JonLiu1993 commented:
@dan-shaw ,Could you please help merge this pr?

mathisloge commented:
Will anyone merge this?

mathisloge committed:
Merge branch 'master' of https://github.com/microsoft/vcpkg into bullet-fix

BillyONeal commented:
Sorry for the delay, thanks for your contribution!
